op-batcher/batcher/driver.go

```diff
@@ -320,3 +327,70 @@ func (l *BatchSubmitter) loop() {
 		}
 	}
 }
+
+func (l *BatchSubmitter) recordFailedTx(id txID, err error) {
+	l.log.Warn("Failed to send transaction", "err", err)
+	l.state.TxFailed(id)
+}
+
+func (l *BatchSubmitter) recordConfirmedTx(id txID, receipt *types.Receipt) {
+	l.log.Info("Transaction confirmed", "tx_hash", receipt.TxHash, "status", receipt.Status, "block_hash", receipt.BlockHash, "block_number", receipt.BlockNumber)
+	// Unfortunately, a tx receipt doesn't include the timestamp, so we have to
+	// query the header.
+	l1ref, err := l.l1BlockRefByReceipt(l.ctx, receipt)
+	if err != nil {
+		// Very unlikely that tx sending worked but then we cannot get the
+		// header. Fall back to latest known L1 time to be on the safe side.
+		l1ref.Time = l.lastKnownL1Time
+		l.log.Warn("Failed to get block ref for successful batcher tx. Setting timestamp to latest know L1 block time.", "block_ref", l1ref)
+	} else {
+		l.lastKnownL1Time = l1ref.Time
+	}
```

Comment:
We could simplify this quite a bit if we measure batcher channel timeouts in #blocks instead of seconds, because block numbers are part of tx receipts. It might then also be enough to only pass a `BlockID` to `TxData` instead of a `L1BlockRef`.

Reply:
done

Reply:
done